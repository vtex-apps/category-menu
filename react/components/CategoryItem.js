import React, { Component } from 'react'
import { Link } from 'render'
import PropTypes from 'prop-types'
import { categoryItemShape } from '../propTypes'

import ItemContainer from './ItemContainer'
import classNames from 'classnames'

/**
 * Component that represents a single category displayed in the menu, also displays
 * the subcategories, if the provided category has them
 */
export default class CategoryItem extends Component {
  state = {
    isHover: false,
  }

  static propTypes = {
    /** Category to be displayed */
    category: categoryItemShape.isRequired,
    /** Set use of Link component */
    noRedirect: PropTypes.bool,
  }

  render() {
    const { category } = this.props
    const { isHover } = this.state

    const containerStyle = {
      top: this.item && this.item.offsetTop + this.item.clientHeight,
      display: isHover ? 'flex' : 'none',
    }

    const linkClasses = classNames(
      'w-100 ph4 pt6 pb5 no-underline f6 outline-0 db tc truncate bb bw1 gray', {
        'b--transparent': !isHover,
        'b--heavy-blue': isHover,
      }
    )

    const itemClasses = 'vtex-category-menu__item flex justify-center items-center'

    return (
      <div className={itemClasses}
        ref={e => { this.item = e }}
        onMouseEnter={() => this.setState({ isHover: true })}
        onMouseLeave={() => this.setState({ isHover: false })}
      >
        {this.props.noRedirect ? (
          <a href="#" className={linkClasses}>
            {category.name.toUpperCase()}
          </a>
        ) : (
          <Link
            page="store/department"
            params={{ department: category.slug }}
            className={linkClasses}
          >
            {category.name.toUpperCase()}
          </Link>
        )}
        {category.children.length > 0 && (
          <div className="absolute w-100 left-0" style={containerStyle}>
            <ItemContainer categories={category.children} parentSlug={category.slug} />
          </div>
        )}
      </div>
    )
  }
}
