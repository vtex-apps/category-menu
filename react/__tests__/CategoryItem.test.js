import React from 'react'
import { shallow } from 'enzyme'

import CategoryItem from '../components/CategoryItem'

describe('CategoryItem component', () => {
  let wrapper

  beforeEach(() => {
    const categoryMock = {
      id: 1,
      name: 'Category',
      href: '#1',
      slug: 'category',
      hasChildren: true,
      children: [
        {
          id: 2,
          name: 'Sub-category',
          href: '#2',
          slug: 'sub-category',
        },
      ],
    }

    wrapper = shallow(<CategoryItem category={categoryMock} />)
  })

  it('should have the correct href', () => {
    expect(wrapper.find({ params: { department: 'category' } }).exists()).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
