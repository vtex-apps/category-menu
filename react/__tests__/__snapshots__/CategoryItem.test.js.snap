// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem component should match snapshot 1`] = `
<div
  className="vtex-category-menu__item ph4"
  data-testid="category-item"
>
  <a
    className="db no-underline ttu"
    href="#1"
  >
    Category
  </a>
  <CategorySubMenu
    subCategories={
      Array [
        Object {
          "href": "#2",
          "id": 2,
          "name": "Sub-category",
          "slug": "sub-category",
        },
      ]
    }
  />
</div>
`;
